File: Data Structures & Algorithms/climbing-stairs/submission-2.java
class Solution {
    int[] memo;
    public int climbStairs(int n) {
        this.memo = new int[n];
        memo[0] = 1;
        if (n > 1) memo[1] = 2;
        return climb(n);
    }

    public int climb(int n) {
        if (memo[n - 1] > 0) return memo[n - 1];
        int res = climb(n - 1) + climb(n - 2);
        memo[n - 1] = res;
        return res;
    }
}

File: Data Structures & Algorithms/coin-change/submission-2.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        memo = {}

        def backtrack(i):
            if i == 0: return 0
            if i < 0: return float('inf')

            if not i in memo:
                memo[i] = 1 + min((backtrack(i - coin) for coin in coins))

            return memo[i]
        
        res = backtrack(amount)
        return res if res != float('inf') else -1
